### PR TITLE
feat(pallet): follow psp22 standard for `decrease_allowance` and `burn`

### PR DESCRIPTION
**Pull request by chungquantin**
Discussion: https://github.com/r0gue-io/pop-node/pull/300#discussion_r1777416957
- For `decrease_allowance`: `Fungibles::decrease_allowance()` saturating_sub the value, it should `checked_sub` and throw an error instead. Hence, we don't have to handle `InsufficientAllowance` on the contract side.
- For `burn`: `InsufficientBalance` case is not returned in the `burn` pallet api if the `value` exceeds the minted value. In `decrease_balance` method of `pallet-assets`, it is also `saturating_sub`.

### CODE REVIEW

File: pallets/api/src/fungibles/tests.rs [hunk 1]
```diff
@@ -272,20 +272,27 @@ fn decrease_allowance_works() {
...
-		assets::create_mint_and_approve(owner, token, owner, value, spender, value);
-		assert_eq!(Assets::allowance(token, &owner, &spender), value);
+		assert_ok!(Assets::thaw_asset(signed(owner), token));
 		// Owner balance is not changed if decreased by zero.
 		assert_eq!(
 			Fungibles::decrease_allowance(signed(owner), token, spender, 0),
 			Ok(Some(WeightInfo::approve(0, 0)).into())
 		);
 		assert_eq!(Assets::allowance(token, &owner, &spender), value);
+		// "Unapproved" error is returned if the current allowance is less than `value`.
```
Comment (Daanvdplas) [inline pallets/api/src/fungibles/tests.rs:291]:
```suggestion
		// "Unapproved" error is returned if the current allowance is less than amount to decrease with.
```

you were using `value * 2`

File: pallets/api/src/fungibles/tests.rs [hunk 2]
```diff
@@ -416,10 +416,17 @@ fn burn_works() {
...
 
-		// Check error works for `Assets::burn()`.
-		assert_noop!(Fungibles::burn(signed(owner), token, from, value), AssetsError::Unknown);
 		assets::create_and_mint_to(owner, token, from, total_supply);
 		assert_eq!(Assets::total_supply(TOKEN), total_supply);
+		// Check error works for `Assets::burn()`.
+		assert_ok!(Assets::freeze_asset(signed(owner), token));
+		assert_noop!(Fungibles::burn(signed(owner), token, from, value), AssetsError::AssetNotLive);
+		assert_ok!(Assets::thaw_asset(signed(owner), token));
+		// "BalanceLow" error is returned if the balance is less than `value`.
```
Comment (Daanvdplas) [inline pallets/api/src/fungibles/tests.rs:425]:
```suggestion
		// "BalanceLow" error is returned if the balance is less than amount to burn.
```

**Review by Daanvdplas — CHANGES_REQUESTED** [review]
Unfortunately still some things to do.
- Why are there so many changes in other files? Makes the PR very messy. 
- Would be nice to cover the changes for the integration tests here as well.
- burn's pre-dispatch weight needs to be changed. Easiest thing to do is adding the benchmark weight for balance read function.
- decrease_allowance needs to be re-benchmarked.

**Comment by chungquantin** [comment]
@Daanvdplas Weird things with Github commit: https://github.com/orgs/community/discussions/6814. Please ignore those empty files.

File: pallets/api/src/fungibles/mod.rs [hunk 3]
```diff
@@ -460,13 +462,17 @@ pub mod pallet {
...
 		pub fn burn(
 			origin: OriginFor<T>,
 			token: TokenIdOf<T>,
 			account: AccountIdOf<T>,
 			value: BalanceOf<T>,
 		) -> DispatchResult {
+			let current_balance = AssetsOf::<T>::balance(token.clone(), &account);
+			if current_balance < value {
+				return Err(AssetsErrorOf::<T>::BalanceLow.into());
+			}
```
Comment (Daanvdplas) [inline pallets/api/src/fungibles/mod.rs:475]: It should only return the weight of querying the balance if it errors

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 4]
```diff
@@ -267,16 +267,21 @@ fn decrease_allowance_works() {
...
-		assert_ok!(decrease_allowance(&addr, 0, &BOB, amount / 2 - 1 * UNIT));
+		assert_ok!(decrease_allowance(&addr, token, &BOB, amount / 2 - 1 * UNIT));
 		let allowance_after = Assets::allowance(token, &addr, &BOB);
 		assert_eq!(allowance_before - allowance_after, amount / 2 - 1 * UNIT);
 		// Token is not live, i.e. frozen or being destroyed.
 		assets::start_destroy(&addr, token);
 		assert_eq!(
 			decrease_allowance(&addr, token, &BOB, amount),
-			Err(Module { index: 52, error: [16, 0] }),
+			Err(Module { index: 52, error: [10, 0] }),
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/fungibles/mod.rs:284]: Same mistake

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 5]
```diff
@@ -256,7 +256,7 @@ fn decrease_allowance_works() {
 		// Token does not exist.
 		assert_eq!(
 			decrease_allowance(&addr, 0, &BOB, amount),
-			Err(Module { index: 52, error: [3, 0] }),
+			Err(Module { index: 52, error: [10, 0] }),
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/fungibles/mod.rs:259]: The comment above this assert is outdated if you simply correct the error code. 

Can you please check your own PRs before you ask for a final review. I have been correcting this same mistake in your other PR as well.

**Review by Daanvdplas — CHANGES_REQUESTED** [review]

File: pallets/api/src/fungibles/mod.rs (hunk 3 above)
Comment (chungquantin) [reply pallets/api/src/fungibles/mod.rs:476]: Resolved!

**Comment by evilrobot-01** [comment]
> @Daanvdplas Weird things with Github commit: https://github.com/orgs/community/discussions/6814. Please ignore those empty files.

Why are the permissions changing? Can we just not revert whatever was changed so that they dont appear within the PR?

File: pallets/api/src/fungibles/tests.rs [hunk 6]
```diff
@@ -416,10 +417,17 @@ fn burn_works() {
 		let from = BOB;
 		let total_supply = value * 2;
 
-		// Check error works for `Assets::burn()`.
-		assert_noop!(Fungibles::burn(signed(owner), token, from, value), AssetsError::Unknown);
```
Comment (evilrobot-01) [inline pallets/api/src/fungibles/tests.rs:420]: Whilst the assets error assertions were to ensure that errors are properly handled, which is the case on line 424, this line could have been left in.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Looks good, but I think the permissions should be reverted to clean up the scope of this PR. Good to merge otherwise.

Comments are only comments.

File: pallets/api/src/fungibles/weights.rs [hunk 7]
```diff
@@ -2,7 +2,7 @@
 //! Autogenerated weights for `fungibles`
 //!
 //! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 40.0.0
-//! DATE: 2024-09-13, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! DATE: 2024-10-01, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/api/src/fungibles/weights.rs:5]: Love your suggestion. Created an issue for this: https://github.com/r0gue-io/pop-node/issues/321

**Comment by chungquantin** [comment]
Created this PR https://github.com/r0gue-io/pop-node/pull/322 to substitute this PR due to the file permission changes. Did try to revert but does not work for me.

**Pull request closed without merging**